mmpose/core/evaluation/top_down_eval.py

```diff
@@ -250,6 +250,83 @@ def _taylor(heatmap, coord):
     return coord
 
 
+def post_dark(coords, batch_heatmaps, share_heatmap=True):
+    """DARK post-pocessing.
+
+    Note:
+        batch_size: N
+        num_keypoints: K
+
+    Args:
+        coords (np.ndarray[N, K, 2]): Coords in shape of batchsize*num_kps*2.
+        batch_heatmaps (np.ndarray[N, K, H, W]): batch_heatmaps
+        share_heatmap (bool) : True for bottom_up paradigm where all persons
+                    share the same heatmap while False for top_down paradigm
+                    where each person has its own heatmaps.
+
+    Returns:
+        res (np.ndarray): Refined coords in shape of batchsize*num_kps*2.
+    """
+    if not isinstance(batch_heatmaps, np.ndarray):
+        batch_heatmaps = batch_heatmaps.cpu().numpy()
+    batch, num_kps, h, w = batch_heatmaps.shape
+    persons = coords.shape[0]
+    if share_heatmap:
+        assert batch == 1
+    else:
+        assert batch == persons
+    for i in range(batch):
+        for j in range(num_kps):
+            mapij = batch_heatmaps[0, j, :, :]
+            mapij = cv2.GaussianBlur(mapij, (3, 3), 0)
+            batch_heatmaps[0, j, :, :] = mapij
+    batch_heatmaps = np.clip(batch_heatmaps, 0.001, 50)
+    batch_heatmaps = np.log(batch_heatmaps)
+    batch_heatmaps = np.transpose(batch_heatmaps, (2, 3, 0, 1)).reshape(
+        (h, w, batch * num_kps))
+    batch_heatmaps_pad = cv2.copyMakeBorder(
+        batch_heatmaps, 1, 1, 1, 1, borderType=cv2.BORDER_REFLECT)
+    batch_heatmaps_pad = np.transpose(
+        batch_heatmaps_pad.reshape((h + 2, w + 2, batch, num_kps)),
+        (2, 3, 0, 1))
+    index = coords[:, :, 0] + 1 + (coords[:, :, 1] + 1) * (w + 2)
+
+    index = index.reshape((persons, -1))
+    if share_heatmap:
+        offset = np.arange(0, num_kps) * (w + 2) * (h + 2)
+        index = index + offset[None, :]
+    else:
+        offset = np.arange(0, batch * num_kps) * (w + 2) * (h + 2)
+        index = index + offset.reshape((persons, -1))
+    index = index.astype(np.int).reshape((-1, 1))
+    batch_heatmaps_pad = batch_heatmaps_pad.reshape((-1, ))
+    i_ = batch_heatmaps_pad[index]
```

Comment:
haven't looked at the details, could the following operations be re-written using conv?

Reply:
yes, re-written using conv may be more elegance but at the cost of more calculation?

Reply:
why? conv should save some computations i guess

Reply:
em, B*K*H*W (conv) vs N*K? 
batch_size: B
num_keypoints: K
num_persons: N
hight_of_heatmaps: H
width_of_heatmaps: W

Reply:
Okay I will check it tonight

Reply:
I have read the codes, the computation is okay.

The only thing to check is that this function will change (when the input is ndarray) the values of batch_heatmaps inplace, so it is kind of unsafe. Please confirm that this would not cause trouble

Reply:
so far, it is save, as it is call at the last ~

Reply:
btw, have you tried to call this function with batch size > 1?

Reply:
you means to use this function in top-down paradigm? yes, i committed the code after i have tested the result of three baseline top-down-GaussianHeatmap/top-down-CombinedTraget/bottom-up-GaussianHeatmap

Reply:
Another question, what if I change the padding to zero padding? Is it supposed to affect results?

Reply:
i test this just now with two different configuration and the results show that this may have precision loss by 0.1~0.2 AP

Reply:
I mean cv2.copyMakeBorder might not support channel number > 3

![image](https://user-images.githubusercontent.com/9464825/101488501-a6bbbb80-399a-11eb-8d1f-3230d16d1805.png)

Reply:
emmm, it works well~ for all configs the channel number is Batch*num_keypoints>3
i use this version opencv~ 
opencv-python          3.4.2.17

Reply:
I see, it errors when bs is large

![image](https://user-images.githubusercontent.com/9464825/101490707-b4267500-399d-11eb-8f22-d3d63e1a1f80.png)

Reply:
Let's leave it as is for now... until someone raise an issue

Reply:
add an assert alone with explanation to avoid this?

Reply:
Simply ignore it is okay

Reply:
ok